src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs

```
@@ -0,0 +1,277 @@
+            _messagePumpSettings = serviceProvider.GetRequiredService<AzureServiceBusMessagePumpSettings>();
+            _messageHandlerOptions = DetermineMessageHandlerOptions(_messagePumpSettings);
+        }
+
+        /// <summary>
+        ///     Path of the entity to process
+        /// </summary>
+        public string EntityPath { get; private set; }
+
+        /// <summary>
+        ///     Service Bus namespace that contains the entity
+        /// </summary>
+        public string Namespace { get; private set; }
+
+        /// <inheritdoc />
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            _messageReceiver = await CreateMessageReceiverAsync(_messagePumpSettings);
+
+            Logger.LogInformation("Starting message pump on entity path '{EntityPath}' in namespace '{Namespace}'",
+                EntityPath, Namespace);
+
+            _messageReceiver.RegisterMessageHandler(HandleMessageAsync, _messageHandlerOptions);
+            Logger.LogInformation("Message pump started");
+
+            await UntilCancelledAsync(stoppingToken);
+
+            Logger.LogInformation("Closing message pump");
+            await _messageReceiver.CloseAsync();
+            Logger.LogInformation("Message pump closed : {Time}", DateTimeOffset.UtcNow);
+        }
+
+        /// <summary>
+        ///     Marks a message as completed
+        /// </summary>
+        /// <remarks>This should only be called if <see cref="AzureServiceBusMessagePumpOptions.AutoComplete" /> is disabled</remarks>
+        /// <param name="lockToken">
+        ///     Token used to lock an individual message for processing. See
+        ///     <see cref="AzureServiceBusMessageContext.LockToken" />
+        /// </param>
+        protected virtual async Task CompleteMessageAsync(string lockToken)
+        {
+            Guard.NotNullOrEmpty(lockToken, nameof(lockToken));
+
+            await _messageReceiver.CompleteAsync(lockToken);
```

Since the `_messageReceiver` is late-assigned; it should be checked before used, no?

---

Although it can practically not be called before it's created it's still a good point, but how would you react to it?

Any kind of exception you are thinking of? Silently proceeding would not be an option if you ask me.

---

It's not only here but further up the file too.
I was thinking of a `InvalidOperationException` because it would happen if the methods were called in an invalid order, hence the object is in an invalid state.

---

Good exception suggestion, added a null check